rfcs/20211025-augru/README.md

```diff
@@ -0,0 +1,1089 @@
+Proposal for GRU with Attentional Update in DNNL
+================================================
+
+## 1. Introduction
+
+In addition to the regular GRU there is a notion of GRU with attentional update
+gate (AUGRU) that can help to model interest evolving in Click-Through Rate
+Prediction [[#1]][1].
+
+
+### 1.1. GRU
+
+GRU is defined by the following formula:
+
+![GRU-LBR](formula-gru-lbr.png)
+
+
+### 1.2. AUGRU
+
+AUGRU is defined by the following formula (the difference from the
+regular GRU is highlighted in red):
+
+![AUGRU-LBR](formula-augru-lbr.png)
+
+
+
+> Warning
+>
+> Attention is passed not as part of source layer as it is done in Neural Machine
+> Translation (NMT) models [[#3]][3], but as a separate memory which is
+> applied to intermediate results (to gate U). This difference makes it impossible
+> to apply attention on application side and pass the result to oneDNN RNN
+> primitive.
+
+## 2. API Changes
+
+Attention memory descriptor can be built from existing `rnn_desc_t` fields (memory
+descriptors, cell kind, direction, propagation kind, etc.). Taking into account
+this fact and to avoid breaking changes an attention memory descriptor will not
+be part of `dnnl_rnn_desc_t`. Instead a user should use query API to get
+attention shapes.
+
+ * Pros:
+   - RNN primitive descriptor can request the best layout for an attention memory
+   - RNN primitive descriptor can request the data type compatible with RNN cell
```

Comment:
That has also the potential of triggering spurious reorders.
In particular, we typically use f32 datatype for intermediate computations in RNN cells, so we would likely return an f32 descriptor. However, the datatype/quantization model used in the framework might lead to attention vectors generated in bf16/int8.

Reply:
Correct, "Cons" contain this bullet. For int8 either framework or oneDNN should dequantize score to apply them to intermediate AUGRU results:
- If it is done on framework side we can get an overhead associated with dequantization
- If it is done on oneDNN side then we are good. During AUGRU creation we could add quantization parameter for attention as it is done for other arguments, so AUGRU will return int8 md.

For bf16 there will be no option for a user to pass an information that attention is bf16, hence we will get an overhead due to conversion bf16->fp32 on framework side.